controllers/vmware/virtualmachinegroup_reconciler.go

```diff
@@ -0,0 +1,400 @@
+// state of VMs belonging to MachineDeployments (handling scale-up/down events).
+
+func (r *VirtualMachineGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the Cluster instance.
+	cluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	log = log.WithValues("Cluster", klog.KObj(cluster))
+
+	// If Cluster is deleted, just return as VirtualMachineGroup will be GCed and no extra processing needed.
+	if !cluster.DeletionTimestamp.IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	// If ControlPlane haven't initialized, requeue it since VSphereMachines of MachineDeployment will only be created after
+	// ControlPlane is initialized.
+	if !conditions.IsTrue(cluster, clusterv1.ClusterControlPlaneInitializedCondition) {
+		log.Info("Waiting for Cluster ControlPlaneInitialized")
+		return reconcile.Result{RequeueAfter: reconciliationDelay}, nil
+	}
+
+	// Continue with the main logic.
+	return r.createOrUpdateVirtualMachineGroup(ctx, cluster)
+}
+
+// createOrUpdateVirtualMachineGroup Create or Update VirtualMachineGroup.
+func (r *VirtualMachineGroupReconciler) createOrUpdateVirtualMachineGroup(ctx context.Context, cluster *clusterv1.Cluster) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Calculate current Machines of all MachineDeployments.
+	current, err := getCurrentVSphereMachines(ctx, r.Client, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "failed to get current VSphereMachine of cluster %s/%s",
+			cluster.Name, cluster.Namespace)
+	}
+
+	desiredVMG := &vmoprv1.VirtualMachineGroup{}
+	key := &client.ObjectKey{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name,
+	}
+
+	if err := r.Client.Get(ctx, *key, desiredVMG); err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.Error(err, "failed to get VirtualMachineGroup")
+			return reconcile.Result{}, err
+		}
+
+		// Calculate expected Machines of all MachineDeployments.
+		// CAPV retrieves placement decisions from the VirtualMachineGroup to guide
+		// day-2 VM placement. At least one VM is required for each MachineDeployment.
+		expected, err := getExpectedVSphereMachineCount(ctx, r.Client, cluster)
+		if err != nil {
+			log.Error(err, "failed to get expected Machines of all MachineDeployment")
+			return reconcile.Result{}, err
+		}
+
+		if expected == 0 {
+			errMsg := fmt.Sprintf("Found 0 desired VSphereMachine for Cluster %s/%s", cluster.Name, cluster.Namespace)
+			log.Error(nil, errMsg)
+			return reconcile.Result{}, errors.New(errMsg)
+		}
+
+		// Wait for all intended VSphereMachines corresponding to MachineDeployment to exist only during initial Cluster creation.
+		current := int32(len(current))
```

Comment:
this is very confusing, we are shadowing a variable that contains a list of virtual machines with another with the same name, but different type and meaning.


```suggestion
		currentVSphereMachineCount := int32(len(current))
```

Reply:
Fixed.